### PR TITLE
Ctests for individual schemes. 

### PR DESCRIPTION
**Pull request by bbakernoaa**
## Description of Changes:
This pull request introduces comprehensive unit testing for the FENGSHA dust emission scheme in the `physics/smoke_dust` directory. It adds a new CTest-based workflow for automated testing, includes a CMake configuration for building the tests, and implements a thorough Fortran test program that covers multiple edge cases for dust emissions. Additionally, a minor error handling improvement is made in the smoke wrapper routine.

**Automated Testing Infrastructure:**

* Adds a new GitHub Actions workflow (`.github/workflows/ci_ctests.yml`) to automatically build and run the FENGSHA unit tests on push and pull request events.
* Introduces a `CMakeLists.txt` file for the FENGSHA test suite, enabling easy compilation and integration with CTest for automated test execution.

**FENGSHA Dust Emission Unit Tests:**

* Implements a comprehensive Fortran test program (`test_dust_fengsha.F90`) that exercises the FENGSHA dust emission driver across a variety of scenarios, including standard emission, water, frozen soil, snow cover, high roughness, invalid soil type, low wind, and very dry soil. The test program reports pass/fail for each case and stops on failure.

**Code Robustness Improvement:**

* Updates the `rrfs_smoke_wrapper_run` subroutine to propagate error messages and flags from the FENGSHA driver and to return early if an error is detected, improving overall error handling.

@climbfuji @ligiabernardet @dustinswales

### CODE REVIEW

File: .github/workflows/ci_ctests.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: Physics Scheme Unit Tests (CTest)
+
+on: [push, pull_request]
+
+jobs:
+  ctests:
+    if: github.repository == 'NCAR/ccpp-physics' || github.repository == 'ufs-community/ccpp-physics' || github.repository == 'bbakernoaa/ccpp-physics'
```
Comment (climbfuji) [inline .github/workflows/ci_ctests.yml:7]: I think this should be either "only the authoritative repository" or "any repository".

File: .github/workflows/ci_ctests.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v3
+
+    - name: Install dependencies
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y gfortran cmake
```
Comment (climbfuji) [inline .github/workflows/ci_ctests.yml:17]: This may be ok for now, in the long run we need another option so that we have ESMF, w3emc, ip, bacio. We can probably piggyback on the container solution that @dustinswales is working on for running SCM from this repository.

File: .github/workflows/ci_ctests.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+on: [push, pull_request]
+
+jobs:
+  ctests:
+    if: github.repository == 'NCAR/ccpp-physics' || github.repository == 'ufs-community/ccpp-physics' || github.repository == 'bbakernoaa/ccpp-physics'
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v3
```
Comment (climbfuji) [inline .github/workflows/ci_ctests.yml]:
```suggestion
      uses: actions/checkout@v4
```
or whatever the latest is

File: physics/smoke_dust/tests/CMakeLists.txt
Comment (climbfuji) [inline physics/smoke_dust/tests/CMakeLists.txt:1]: We'll need a bit of time to mull this over. It may be better to have one `test` directory at the top level of NCAR ccpp-physics main that takes care of the different tests, like it is done for most other large code bases. (I am talking about the CMakeLists.txt part only).

File: .github/workflows/ci_ctests.yml (hunk 1 above)
Comment (bbakernoaa) [reply .github/workflows/ci_ctests.yml:7]: I agree. These are small enough it could any repository.

File: .github/workflows/ci_ctests.yml (hunk 2 above)
Comment (bbakernoaa) [reply .github/workflows/ci_ctests.yml:17]: Agreed

File: physics/smoke_dust/tests/CMakeLists.txt
Comment (bbakernoaa) [reply physics/smoke_dust/tests/CMakeLists.txt:1]: Sure. Just a jumping off point.

Comment (dustinswales) [reply physics/smoke_dust/tests/CMakeLists.txt:1]: I kinda like the idea of keeping the tests alongside the scheme, which is consistent with how physics submodules are organized (i.e. test/ and src/ dirs).

File: .github/workflows/ci_ctests.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+name: Physics Scheme Unit Tests (CTest)
```
Comment (dustinswales) [inline .github/workflows/ci_ctests.yml:1]: Maybe change to something like "Smoke-Dust Unit Tests (CTest)"

File: .github/workflows/ci_ctests.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+on: [push, pull_request]
+
+jobs:
+  ctests:
```
Comment (dustinswales) [inline .github/workflows/ci_ctests.yml:6]: Ditto more descriptive name here.
If/when more schemes add unit tests, we want to have a name that is extensible

File: .github/workflows/ci_ctests.yml (hunk 2 above)
Comment (dustinswales) [reply .github/workflows/ci_ctests.yml:17]: These unit tests don't need ESMF or any other dependencies, just fortran and cmake.

The CCPP SCM tests do need all of those other things, which is a pain to install/maintain on the GitHub runners, not to mention slow. So we are [moving to containerized CI](https://github.com/NCAR/ccpp-scm/pull/658) to give us total control, plus access to proprietary compilers.

**Review by dustinswales — CHANGES_REQUESTED** [review]
@bbakernoaa Thanks for these changes.
You are the first developer to add unit tests to a scheme in the CCPP and trigger them via GitHub Actions. Schemes in submodules do this, but no parameterizations residing in the CCPP do.

We need more of this in the CCPP and hopefully this provides the motivation and a roadmap for others.

I suggest using more specific names for the tests, but otherwise I have no concerns.

File: .github/workflows/ci_ctests.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+name: Physics Scheme Unit Tests (CTest)
+
+on: [push, pull_request]
```
Comment (grantfirl) [inline .github/workflows/ci_ctests.yml:3]: How about this is triggered using the following:
on:
  push:
    branches: [main]
  pull_request:
  workflow_dispatch:

Comment (grantfirl) [reply .github/workflows/ci_ctests.yml:3]: I'm not sure that we want this to run on every push to save GitHub resources.

File: .github/workflows/ci_ctests.yml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        sudo apt-get install -y gfortran cmake
+
+    - name: Build and run smoke and dust tests
+      run: |
+        cd physics/smoke_dust/tests
+        mkdir build
+        cd build
+        cmake ..
+        make
+        ctest --output-on-failure
```
Comment (rhaesung) [inline .github/workflows/ci_ctests.yml:26]: To make debugging easier when a test fails in the CI, would it be possible to add an `upload-artifact` step?

File: physics/smoke_dust/tests/test_dust_fengsha.F90
Comment (rhaesung) [inline physics/smoke_dust/tests/test_dust_fengsha.F90:1]: The variety of scenarios covered here is great. Are there any plans to add numerical value checks to these tests in the future? It might be a nice secondary filter before we hit the full UFS RTs?

Comment (bbakernoaa) [reply physics/smoke_dust/tests/test_dust_fengsha.F90:1]: Yes we should have numerical value checks as well.  This was meant as a proof of concept.  They should of course be expanded

File: .github/workflows/ci_ctests.yml (hunk 7 above)
Comment (bbakernoaa) [reply .github/workflows/ci_ctests.yml:26]: yes we could do that

File: .github/workflows/ci_ctests.yml (hunk 6 above)
Comment (bbakernoaa) [reply .github/workflows/ci_ctests.yml:3]: these are very lightweight and wouldn't really be a limiting factor in my opinion.  We should do it on PRs so that we know within the PR if it is failing before the PR is approved.

File: .github/workflows/ci_ctests.yml (hunk 4 above)
Comment (bbakernoaa) [reply .github/workflows/ci_ctests.yml:1]: sure

**Comment by rhaesung** [comment]
@bbakernoaa Is this PR waiting for more updates?

**Comment by bbakernoaa** [comment]
> @bbakernoaa Is this PR waiting for more updates?

It still needs a little work.  I'll change it to a draft as I'm pretty busy with things right now

**Comment by climbfuji** [comment]
> @bbakernoaa Is this PR waiting for more updates?

As far as I am concerned, this PR was submitted as a pilot/proof of concept and we want to take a bit of time to design the automated testing hardness for ccpp-physics.

**Comment by grantfirl** [comment]
I added an issue in the authoritative repo (https://github.com/NCAR/ccpp-physics/issues/1238) and copied this branch over to the NCAR fork whenever someone has time to pick this up. I'm closing this PR to clean up our PR list.
